sw/device/lib/dif/dif_otp_ctrl.h

```diff
@@ -0,0 +1,957 @@
+   *
+   * A followup read should work as expected.
+   */
+  kDifOtpCtrlErrorRecoverableRead,
+  /**
+   * Indicates an unrecoverable error during a read operation.
+   *
+   * This error indicates non-recoverable hardware malfunction.
+   */
+  kDifOtpCtrlErrorUnrecoverableRead,
+  /**
+   * Indicates an unspecified error during a read operation.
+   *
+   * This error indicates non-recoverable hardware malfunction.
+   */
+  kDifOtpCtrlErrorUnspecifiedRead,
+  /**
+   * Indicates that the blank write check failed during a write operation.
+   */
+  kDifOtpCtrlErrorBlankCheckFailed,
+  /**
+   * Indicates an unspecified error during a write operation.
+   *
+   * This error indicates non-recoverable hardware malfunction.
+   */
+  kDifOtpCtrlErrorUnspecifiedWrite,
+  /**
+   * Indicates an invalid command was issued to the DAI.
+   */
+  kDifOtpCtrlErrorInvalidDaiCommand,
+  /**
+   * Indicates a locked memory region was accessed.
+   */
+  kDifOtpCtrlErrorLockedAccess,
+  /**
+   * Indicates an integrity check failed in the buffer registers.
+   *
+   * This error indicates non-recoverable hardware malfunction.
+   */
+  kDifOtpCtrlErrorIntegrityLoss,
+  /**
+   * Indicates a consistency check failed in the buffer registers.
+   *
+   * This error indicates non-recoverable hardware malfunction.
+   */
+  kDifOtpCtrlErrorConsistencyLoss,
+  /**
+   * Indicates that the FSM of the controller is in a bad state.
+   *
+   * This error indicates that the device has been glitched by an attacker.
+   */
+  kDifOtpCtrlErrorFsmBadState,
+  /**
+   * Indicates that an escalation occured in the alert subsystem, and the
+   * controller's FSM is in its terminal state.
+   */
+  kDifOtpCtrlErrorFsmTerminalState,
+} dif_otp_ctrl_error_t;
+
+/**
+ * The overall status of the OTP controller.
+ *
+ * See `dif_otp_ctrl_get_status()`.
+ */
+typedef struct dif_otp_ctrl_status {
+  /**
+   * Currently active statuses, given as a bit vector. To check whether a
+   * particular status code was returned, write
+   *
+   *   bool has_code = (status.codes >> kMyStatusCode) & 1;
+   *
+   * Note that it is possible to quickly check that the controller is idle and
+   * error-free by writing
+   *
+   *   bool is_ok = status.codes == (1 << kDifOtpStatusCodeDaiIdle);
+   */
+  uint32_t codes;
+  /**
+   * A list of root causes for each error status code.
+   *
+   * If the error status code `error` is present in `codes`, then
+   * `causes[error]` will contain its root cause.
+   */
+  // Open Q: This array is much wider than it needs to be, since not all status
+  // codes have associated errors. We could make this array shorter, since the
+  // statuses that *can* have errors are the prefix of the enum. Thoughts?
+  dif_otp_ctrl_error_t causes[kDifOtpCtrlStatusCodesCount];
+} dif_otp_ctrl_status_t;
+
+/**
+ * Device lifecycle state stored in OTP memory.
+ *
+ * See `dif_otp_ctrl_get_lifecycle_state()`.
+ */
+typedef struct dif_otp_ctrl_lifecycle_state {
```

Comment:
FYI we decided to remove the raw `LC_STATE` and `LC_TRANSITION_CNT` from OTP (in https://github.com/lowRISC/opentitan/pull/3720) and put them into the life cycle controller CSRs where they belong. so the associated functions to read these CSRs from OTP can be removed.

Reply:
Cool!

Reply:
@mcy can you file a followup issue for this DIF that has this in it?